ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift

```diff
@@ -347,18 +347,40 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
       perform(#selector(self.enableInputClickSound), with: nil, afterDelay: 0.1)
     }
 
-    var hasDeletedSelection = false
+    var deleteSelection = false
 
     if let selected = textDocumentProxy.selectedText {
       if selected.count > 0 {
-        textDocumentProxy.deleteBackward()
-        hasDeletedSelection = true
+        deleteSelection = true
       }
     }
 
-    if numCharsToDelete <= 0 || hasDeletedSelection {
+    if deleteSelection && newText == "" {
+      /*
+        if deleteSelection && newText == "", we have a backspace on
+        selected text.  Sadly, .insertText("")... does nothing.  Why, Apple!?
+       
+        The one silver lining:  Apple makes it impossible for users to select text
+        in a way that splits character clusters.
+       
+        So, we can just insert something that won't combine, like a ZWNJ, and then delete it.
```

Comment:
Are you saying that you are working around the fact that insertText of the empty string does not cause the selection to be replaced by doing an insert of something invisible that we can then delete? Does this only happen for backspace?

It sounds like you are detecting this backspace case (based on line 360) by saying we have an empty string to insert and the text selection is greater than zero. Is that only true for backspace? 

This comment is hard to follow because the problem is identified, then the silver lining is mentioned and then the workaround. Also, it would be nice to keep the emotion or sarcasm out of the code as it can distract from the technical content -- even if the technical content is correct. The same is true for comments in issues or PRs. This is not productive for our relationships with our vendors.

Reply:
> Are you saying that you are working around the fact that insertText of the empty string does not cause the selection to be replaced by doing an insert of something invisible that we can then delete?

Yes.  That's exactly what it's saying.

> Does this only happen for backspace?
>
> It sounds like you are detecting this backspace case (based on line 360) by saying we have an empty string to insert and the text selection is greater than zero. Is that only true for backspace?

It's theoretically possible for other backspace-like keys to be defined, I suppose.  Changes to keyboard options don't trigger `insertText` unless they also _actually_ manipulate text.

Suppose the following:

```swift
self.insertText(keymanWeb , /*numCharsToDelete:*/ 2, /*newText*/ "a")
```

If text is selected that immediately follows a space, we need to _preserve_ that space in order to properly match how many characters the internal Web engine has said to delete.  (In fact... we need to preserve it regardless in order to stay synch'd with the internal engine.)  We have to do the deletions _before_ our text-insertions, too.  

So... I think this needs to happen this way _regardless_ of whether or not we're inserting text down the line - otherwise, we run the risk of desynchronizing.  If `temptext` is the context and `text` is selected, as this code currently is, at the time of this comment, the first char to delete will be silenced due to erasing the selected text while _not_ deleting the preceding `p`.

So... thanks for raising the point that helped me catch this.

Reply:
> Also, it would be nice to keep the emotion or sarcasm out of the code as it can distract from the technical content -- even if the technical content is correct.

I have at times been guilty of this. It's not a good practice, I agree. Frustration can be expressed in Slack ;-) and we'll keep the code comments, issues, and PRs focused on technical.

> So... I think this needs to happen this way _regardless_ of whether or not we're inserting text down the line - otherwise, we run the risk of desynchronizing. If `temptext` is the context and `text` is selected, as this code currently is, at the time of this comment, the first char to delete will be silenced due to erasing the selected text while _not_ deleting the preceding `p`.

.insertText() will delete selection if the string is not empty, right? And in that scenario I don't think it deletes the space prior to the selection. AFAICT, we only need to jump through the hoop when we are deleting selection with no insertion (generally, backspace)